packages/astro/src/runtime/server/index.ts

```diff
@@ -306,32 +317,42 @@ export function createAstro(filePathname: string, site: string, projectRootStr:
 	};
 }
 
-const toAttributeString = (value: any) => String(value).replace(/&/g, '&#38;').replace(/"/g, '&#34;');
+const toAttributeString = (value: any): string => String(value).replace(/&/g, '&#38;').replace(/"/g, '&#34;');
 
 // A helper used to turn expressions into attribute key/value
-export function addAttribute(value: any, key: string) {
-	if (value == null || value === false) {
-		return '';
-	}
-
+export async function addAttribute(value: any, key: string): Promise<string> {
 	// support "class" from an expression passed into an element (#782)
 	if (key === 'class:list') {
 		return ` ${key.slice(0, -5)}="${toAttributeString(serializeListValue(value))}"`;
 	}
 
-	// Boolean only needs the key
-	if (value === true && key.startsWith('data-')) {
-		return ` ${key}`;
-	} else {
-		return ` ${key}="${toAttributeString(value)}"`;
+	const attributeValue = await getAttributeValue(value)
+
+	return attributeValue === '' ? ` ${key}` : ` ${key}="${attributeValue}"`
+}
+
+// A helper used to turn expressions into attribute key/value
+export async function getAttributeValue(value: any): Promise<string> {
+	switch (typeOf(value)) {
+		case 'null':
+		case 'undefined':
+		case value === '' && 'string':
+		case value === false && 'boolean':
+			return ''
+		case 'iterable':
+			return (await Promise.all(Array.from(value, getAttributeValue))).join('')
```

Comment:
Possibly be explicit in the `Array.from` callback to avoid leaking the second `index` argument?
 
```suggestion
			return (await Promise.all(Array.from(value, (value) => getAttributeValue(value)))).join('')
```